### PR TITLE
Kafka reference guide

### PR DESCRIPTION
**Pull request by ozangunalp**
Transform Getting started to Reference Guide

- Remove example application to a separate getting started guide: https://github.com/quarkusio/quarkus/pull/17830
- Receiving messages: method signatures, acknowledgments & commit strategies, error handling, consumer groups, blocking processing, consumer rebalance listener
- Sending messages: method signatures, message acknowledgment, backpressure, retries, in-memory channels, broadcasting
- Processing messages
- Accessing Kafka clients directly
- Serialization JSON
- Health Checks
- Message compression
- Authentication 
- Testing applications
- Kubernetes service binding
- Configuration reference

Links to other guides : Avro Serialization & Schema Registry, Kafka Streams, Dev Services

### CODE REVIEW

File: docs/src/main/asciidoc/kafka.adoc [hunk 1]
```diff
@@ -3,438 +3,946 @@ This guide is maintained in the main Quarkus repository
 and pull requests should be submitted there:
 https://github.com/quarkusio/quarkus/tree/main/docs/src/main/asciidoc
 ////
-= Quarkus - Using Apache Kafka with Reactive Messaging
+= Quarkus - Apache Kafka Reference Guide
```
Comment (cescoffier) [inline docs/src/main/asciidoc/kafka.adoc:6]:
```suggestion
= Apache Kafka Reference Guide
```

File: docs/src/main/asciidoc/kafka.adoc [hunk 2]
```diff
@@ -3,438 +3,946 @@ This guide is maintained in the main Quarkus repository
...
+[source, properties]
 ----
-mvn io.quarkus:quarkus-maven-plugin:{quarkus-version}:create \
-    -DprojectGroupId=org.acme \
-    -DprojectArtifactId=kafka-quickstart \
-    -DclassName="org.acme.kafka.PriceResource" \
-    -Dpath="/prices" \
-    -Dextensions="resteasy-reactive,smallrye-reactive-messaging-kafka"
-cd kafka-quickstart
+kafka.bootstrap.servers=kafka:9092 <1>
```
Comment (cescoffier) [inline docs/src/main/asciidoc/kafka.adoc:63]: I would remove it, or use `%prod.kafka.bootstrap.servers`, as we want to encourage the usage of dev services.

File: docs/src/main/asciidoc/kafka.adoc [hunk 3]
```diff
@@ -3,438 +3,946 @@ This guide is maintained in the main Quarkus repository
...
+mp.messaging.incoming.prices.connector=smallrye-kafka <2>
 ----
 
-This command generates a Maven project, importing the Reactive Messaging and Kafka connector extensions.
+<1> Configure the broker location. You can configure it globally or per channel. It defaults to localhost:9092
+<2> Configure the connector to manage the prices channel. By default the topic name is same as the channel name. You can configure the topic attribute to override it.
 
-If you already have your Quarkus project configured, you can add the `smallrye-reactive-messaging-kafka` extension
-to your project by running the following command in your project base directory:
+== Receiving messages from Kafka
```
Comment (cescoffier) [inline docs/src/main/asciidoc/kafka.adoc:70]: If you remove the bootstrap servers below add a note here indicating that is can be configured globally or per channel and mention dev services.

File: docs/src/main/asciidoc/kafka.adoc [hunk 4]
```diff
@@ -3,438 +3,946 @@ This guide is maintained in the main Quarkus repository
...
+----
 
-import java.time.Duration;
-import java.util.Random;
+`Record` is a simple wrapper around key and payload of the incoming Kafka record.
 
-import javax.enterprise.context.ApplicationScoped;
+.@Channel
 
+// I would show an example with Multi and not raw publisher.
```
Comment (cescoffier) [inline docs/src/main/asciidoc/kafka.adoc:142]: Remove comment

File: docs/src/main/asciidoc/kafka.adoc [hunk 5]
```diff
@@ -3,438 +3,946 @@ This guide is maintained in the main Quarkus repository
...
+Alternatively, your application can inject a Reactive Stream `Publisher` in your bean and subscribe to its events as the following example:
```
Comment (cescoffier) [inline docs/src/main/asciidoc/kafka.adoc:143]:
```suggestion
Alternatively, your application can inject a `Multi` in your bean and subscribe to its events as the following example:
```

File: docs/src/main/asciidoc/kafka.adoc [hunk 6]
```diff
@@ -3,438 +3,946 @@ This guide is maintained in the main Quarkus repository
...
+The second one, used also with other reactive features of Quarkus, uses the default worker pool and preserves the order.
+====
+
+=== Acknowledgment Strategies
+
+All messages received by a consumer must be acknowledged.
+In the absence of acknowledgment, the processing is considered on error.
+If the consumer method receives a `Record` or a payload, the message will be acked on method return, also known as `Strategy.POST_PROCESSING`.
+If the consumer method returns another reactive stream or `CompletableStage`, the message will be acked when the downstream message is acked.
+// No, it will be ack then the produced message get acked -> Needs confirmation
```
Comment (cescoffier) [inline docs/src/main/asciidoc/kafka.adoc:257]: To be removed.

File: docs/src/main/asciidoc/kafka.adoc [hunk 7]
```diff
@@ -3,438 +3,946 @@ This guide is maintained in the main Quarkus repository
...
+
+The connector only sends that amount of messages concurrently.
+No other messages will be sent until at least one in-flight message gets acknowledged by the broker.
+Then, the connector writes a new message to Kafka when one of the broker’s in-flight messages get acknowledged.
+Be sure to configure Kafka’s `batch.size` and `linger.ms` accordingly.
+
+You can also remove the limit of in-flight messages by setting `max-inflight-messages` to `0`.
+However, note that the Kafka Producer may block if the number of requests reaches `max.in.flight.requests.per.connection`.
+
+=== Retries
```
Comment (cescoffier) [inline docs/src/main/asciidoc/kafka.adoc:745]: Should it be under the producing message section?

File: docs/src/main/asciidoc/kafka.adoc [hunk 8]
```diff
@@ -3,438 +3,946 @@ This guide is maintained in the main Quarkus repository
...
+
+When the Kafka producer receives an error from the server, if it is a transient, recoverable error, the client will retry sending the batch of messages.
+This behavior is controlled by `retries` and `retry.backoff.ms` parameters.
+In addition to this, Smallrye reactive messaging will retry individual messages on recoverable errors, depending on the `retries` and `delivery.timeout.ms` parameters.
+
+Note that while having retries in a reliable system is a best practice, the `max.in.flights.requests.per.session` parameter defaults to `5`, meaning that the order of the messages is not guaranteed.
+If the message order is a must for your use case, setting `max.in.flights.requests.per.session` to `1` will make sure a single batch of messages is sent at a time, in the expense of limiting the throughput of the producer.
+// should we have an option to disable message level retries ?
```
Comment (cescoffier) [inline docs/src/main/asciidoc/kafka.adoc:753]: to be removed.

File: docs/src/main/asciidoc/kafka.adoc [hunk 9]
```diff
@@ -3,438 +3,946 @@ This guide is maintained in the main Quarkus repository
...
+
+=== In-memory channels
+// That should be in the reactive messaging reference guide, but because it does not exist, let's keep it.
```
Comment (cescoffier) [inline docs/src/main/asciidoc/kafka.adoc:756]: to be removed.

File: docs/src/main/asciidoc/kafka.adoc [hunk 10]
```diff
@@ -3,438 +3,946 @@ This guide is maintained in the main Quarkus repository
...
+    public CompletionStage<Void> consume(IncomingKafkaRecord<Integer, String> message) {
+        // We don't need to ACK messages because in this example we set offset during consumer re-balance
+        return CompletableFuture.completedFuture(null);
+    }
+
+}
+----
+
+To configure the inbound connector to use the provided listener we either set the consumer rebalance listener’s identifier:
+// Me thinking that .name is not great anymore. we may want to enhance this with .idenfier.
```
Comment (cescoffier) [inline docs/src/main/asciidoc/kafka.adoc:493]: to be removed - can you create an issue for this?

**Review by cescoffier — CHANGES_REQUESTED** [review]

File: docs/src/main/asciidoc/kafka.adoc (hunk 10 above)
Comment (ozangunalp) [reply docs/src/main/asciidoc/kafka.adoc:493]: 👍

File: docs/src/main/asciidoc/kafka.adoc (hunk 6 above)
Comment (ozangunalp) [reply docs/src/main/asciidoc/kafka.adoc:257]: 👍

File: docs/src/main/asciidoc/kafka.adoc (hunk 7 above)
Comment (ozangunalp) [reply docs/src/main/asciidoc/kafka.adoc:745]: I think we can rename this section to "Retrying message dispatch" and keep it here. I was thinking of adding the "Retrying processing" section near receiving messages/error handling. What do you think ?

File: docs/src/main/asciidoc/kafka.adoc [hunk 11]
```diff
@@ -3,438 +3,980 @@ This guide is maintained in the main Quarkus repository
...
+[source, properties]
 ----
-mvn io.quarkus:quarkus-maven-plugin:{quarkus-version}:create \
-    -DprojectGroupId=org.acme \
-    -DprojectArtifactId=kafka-quickstart \
-    -DclassName="org.acme.kafka.PriceResource" \
-    -Dpath="/prices" \
-    -Dextensions="resteasy-reactive,smallrye-reactive-messaging-kafka"
-cd kafka-quickstart
+%prod.kafka.bootstrap.servers=kafka:9092 <1>
```
Comment (cescoffier) [inline docs/src/main/asciidoc/kafka.adoc:63]: Add a note about the %prod prefix. Something like:

NOTE: The `%prod` prefix indicates that the property is only used when the application runs in _prod_ mode (so not in dev or test). Refer to the xref:config-reference.adoc#profiles[Profile documentation] for further details.

**Review by cescoffier — CHANGES_REQUESTED** [review]
Almost there! Two remaining comments.

File: docs/src/main/asciidoc/kafka.adoc [hunk 12]
```diff
@@ -3,438 +3,980 @@ This guide is maintained in the main Quarkus repository
...
+}
 ----
 
-== Imperative usage
-
-Sometimes, you need to have an imperative way of sending messages.
-
-For example, if you need to send a message to a stream, from inside a REST endpoint, when receiving a POST request.
-In this case, you cannot use `@Outgoing` because your method has parameters.
+== Accessing Kafka clients directly
```
Comment (cescoffier) [inline docs/src/main/asciidoc/kafka.adoc:903]: Add an anchor as this should be referenced from the getting started guide.

File: docs/src/main/asciidoc/kafka.adoc (hunk 11 above)
Comment (ozangunalp) [reply docs/src/main/asciidoc/kafka.adoc:69]: 👍  done

File: docs/src/main/asciidoc/kafka.adoc (hunk 12 above)
Comment (ozangunalp) [reply docs/src/main/asciidoc/kafka.adoc:925]: Added the reference to the getting started guide

**Review by cescoffier — APPROVED** [review]

**Comment by cescoffier** [comment]
@ozangunalp can you squash your commits?

**Comment by ozangunalp** [comment]
@cescoffier done

**Pull request merged**